sympy/series/formal.py

```diff
@@ -812,6 +813,43 @@ def _compute_fps(f, x, x0, dir, hyper, order, rational, full):
 
     result = None
 
+    fps_sym = False
+    terms = []
+    if isinstance(f, Add):
+        for t in Add.make_args(f):
+            terms.append(t)
+    else:
+        terms.append(f)
+
+    nterms = []
+    symb, res = S.Zero, S.Zero
+    for t in terms:
+        if isinstance(t, Mul):
+            nterm = S.One
+            for term in Mul.make_args(t):
+                if isinstance(term, Pow):
+                    if term.exp.is_symbol:
+                        symb = term.exp
+                        fps_sym = True
+                    elif sympify(term.exp).is_integer:
+                            nterm *= x**(term.exp)
+                    else:
+                        for subt in Add.make_args(term.exp):
+                            if subt.is_symbol:
+                                res = term.exp - subt
+                                if sympify(res).is_integer:
+                                    nterm *= x**res
+                                symb = subt
+                                fps_sym = True
+                else:
+                    if not (isinstance(term, Pow) and term.exp.is_symbol):
+                        nterm *= term
+            nterms.append(nterm)
+        else:
+            nterms.append(t)
+
+    f = Add(*nterms)
```

Comment:
It will be helpful if you can add some comments explaining the not so obvious parts. It will help in maintaining it. I feel the code is a little complicated and could be simplified. Is it possible to extract some terms and use the existing algorithms on the remaining terms and maybe get the results back together? Is something like that possible?

Reply:
Actually, the code is all about extracting the symbolic terms, and its various versions, like `x**(n-2)` or `x**(n/2)`, and then applying the rest of the functions (containing only function of x terms) to the existing algorithm. I will try to refactor the code, and add comments wherever necessary.